src/internal/rest.rs

```diff
@@ -559,93 +560,87 @@ impl<'a> IronCoreRequest<'a> {
+            );
+
+            // add query params
+            if let Some(query) = maybe_query_params {
+                Self::req_add_query(req.borrow_mut(), &query);
+            }
 
-                (req, body_bytes)
-            })
-        };
-        make_req()
-            .into_future()
-            .and_then(move |(mut req, body_bytes)| {
-                // use the completed request to finish authorization v2 headers
-                let maybe_auth = SignatureUrlString::new(req.url().as_str())
-                    .map(|sig_url| {
-                        auth_b.finish_with(sig_url, req.method().clone(), Some(&body_bytes))
-                    })
-                    .map_err(|e| IronOxideErr::from((e, error_code)));
-
-                // we only support Authorization::Version2 with this call
-                match maybe_auth {
-                    Ok(auth) => {
-                        if let Authorization::Version2 {
-                            user_context,
-                            request_sig,
-                        } = &auth
-                        {
-                            match user_context.to_header(error_code) {
-                                Ok(user_context_header) => {
-                                    replace_headers(req.headers_mut(), user_context_header);
-                                    replace_headers(req.headers_mut(), DEFAULT_HEADERS.clone());
-                                    replace_headers(req.headers_mut(), auth.to_auth_header());
-                                    replace_headers(req.headers_mut(), request_sig.to_header());
-
-                                    Either::A(Self::send_req(req, error_code, resp_handler))
-                                }
-                                Err(e) => Either::B(futures::future::err(e)),
-                            }
-                        } else {
-                            panic!("authorized requests must use version 2 of API authentication")
+            // add the body
+            let body_bytes: Vec<u8> = if let Some(json_se) = maybe_body {
+                let body = serde_json::to_vec(&json_se)
+                    .map_err(|e| IronOxideErr::from((e, error_code)))?;
+                req.headers_mut()
+                    .insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+                *req.body_mut() = Some(body.clone().into());
+                body
+            } else {
+                vec![]
+            };
+
+            (req, body_bytes)
+        });
+        let (mut req, body_bytes) = make_req?;
+
+        // use the completed request to finish authorization v2 headers
+        let maybe_auth = SignatureUrlString::new(req.url().as_str())
+            .map(|sig_url| auth_b.finish_with(sig_url, req.method().clone(), Some(&body_bytes)))
+            .map_err(|e| IronOxideErr::from((e, error_code)));
+
+        // we only support Authorization::Version2 with this call
```

Comment:
There was some hope that the duplicated code between here and `post_raw` might be able to be easily refactored once we got rid of v1 futures. Is that worth taking a pass at, do you think?